junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/DynamicContainerTestDescriptor.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2015-2017 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v1.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.junit.jupiter.engine.descriptor;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.junit.jupiter.api.DynamicContainer;
+import org.junit.jupiter.api.DynamicNode;
+import org.junit.jupiter.engine.execution.JupiterEngineExecutionContext;
+import org.junit.platform.engine.TestDescriptor;
+import org.junit.platform.engine.TestSource;
+import org.junit.platform.engine.UniqueId;
+
+/**
+ * {@link TestDescriptor} for a {@link DynamicContainer}.
+ *
+ * @since 5.0
+ */
+class DynamicContainerTestDescriptor extends JupiterTestDescriptor {
+
+	private final DynamicContainer dynamicContainer;
+	private final AtomicInteger index;
+
+	DynamicContainerTestDescriptor(UniqueId uniqueId, DynamicContainer dynamicContainer, AtomicInteger index,
+			TestSource source) {
+		super(uniqueId, dynamicContainer.getDisplayName());
+		this.dynamicContainer = dynamicContainer;
+		this.index = index;
+		setSource(source);
+	}
+
+	@Override
+	public Type getType() {
+		return Type.CONTAINER;
+	}
+
+	@Override
+	public boolean isLeaf() {
+		return true;
+	}
+
+	@Override
+	public JupiterEngineExecutionContext execute(JupiterEngineExecutionContext context,
+			DynamicTestExecutor dynamicTestExecutor) throws Exception {
+		TestSource source = getSource().orElseThrow(AssertionError::new);
+		for (DynamicNode childNode : dynamicContainer.getDynamicNodes()) {
```

Comment:
I think every level of the hierarchy should have separate indices starting with 1 on every level.